tests/units/units.cpp

```diff
@@ -2004,3 +2145,111 @@ TEST(Units, scalingFactorAcrossStandardUnits)
     EXPECT_EQ(0.001, libcellml::Units::scalingFactor(u1, u2));
     EXPECT_EQ(1000.0, libcellml::Units::scalingFactor(u2, u1));
 }
+
+TEST(Units, scalingFactorChildUnitsNotFound)
+{
+    auto model = libcellml::Model::create("nurseryrhymes");
+    auto u1 = libcellml::Units::create("bo_peep");
+    u1->addUnit("sheep");
+
+    auto u2 = libcellml::Units::create("flock");
+
+    model->addUnits(u1);
+    model->addUnits(u2);
+
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u1, u2));
+}
+
+TEST(Units, scalingFactorGrandchildUnitsNotFound)
+{
+    auto model = libcellml::Model::create("nurseryrhymes");
+    auto u0 = libcellml::Units::create("sheep");
+    u0->addUnit("wool");
+
+    auto u1 = libcellml::Units::create("bo_peep");
+    u1->addUnit("sheep");
+
+    auto u2 = libcellml::Units::create("flock");
+
+    model->addUnits(u0);
+    model->addUnits(u1);
+    model->addUnits(u2);
+    auto s = libcellml::Units::scalingFactor(u1, u2);
+
+    EXPECT_EQ(0.0, s);
+}
+
+TEST(Units, compatibleUnitsInDifferentModelsChildNotFound)
+{
+    auto m1 = libcellml::Model::create("nurseryrhymes");
+    auto m2 = libcellml::Model::create("farmwork");
+    auto u1 = libcellml::Units::create("bo_peep");
+    u1->addUnit("sheep");
+
+    auto u2 = libcellml::Units::create("flock");
+    auto u3 = libcellml::Units::create("sheep");
+    u2->addUnit("sheep");
+
+    m1->addUnits(u1);
+    m2->addUnits(u2);
+    m2->addUnits(u3);
+
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u1, u2));
+}
+
+TEST(Units, compatibleGrandchildUnitsImported)
+{
+    auto model = libcellml::Model::create("nurseryrhymes");
+
+    auto imported_units = libcellml::Units::create("imported_units");
+    auto imp = libcellml::ImportSource::create();
+    imp->setUrl("i_am_a_url");
+    imported_units->setImportSource(imp);
+
+    auto u0 = libcellml::Units::create("u0");
+    u0->addUnit("imported_units");
+
+    auto u1 = libcellml::Units::create("u1");
+    u1->addUnit("u0");
+
+    auto u2 = libcellml::Units::create("u2");
+    u2->addUnit("u1");
+
+    auto u3 = libcellml::Units::create("u3");
+    u3->addUnit("imported_units");
+
+    model->addUnits(u0);
+    model->addUnits(u1);
+    model->addUnits(u2);
+    model->addUnits(u3);
+    model->addUnits(imported_units);
+
+    // Expect false only because the first units "imported_units" is imported... otherwise it would be true.
+    EXPECT_FALSE(libcellml::Units::compatible(u2, u3));
+}
+
+TEST(Units, scalingFactorBetweenExponentialUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    u1->setName("u1");
+    u1->addUnit("metre", 0, 3.0, 1000.0);
+    libcellml::UnitsPtr u2 = libcellml::Units::create();
+    u2->setName("u2");
+    u2->addUnit("metre", 0, 1.0, 1.0);
+    u2->addUnit("metre", 0, 1.0, 1.0);
+    u2->addUnit("metre", 0, 1.0, 1.0);
+
+    EXPECT_EQ(1000.0, libcellml::Units::scalingFactor(u2, u1));
```

Comment:
Might want to test the scaling factor between `u1` and `u2` too.

Reply:
Done.